components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/util/RuleBuilder.java

```diff
@@ -154,37 +155,58 @@ private void addORCombinedRule() {
         orRuleCount++;
     }
 
-    private void validateExpression(Expression expression) {
+    private Expression validateExpression(Expression expression) {
 
         FieldDefinition fieldDefinition = expressionMetadataFieldsMap.get(expression.getField());
 
         if (isError) {
-            return;
+            return expression;
         }
 
-        if (fieldDefinition == null) {
-            setValidationError("Field " + expression.getField() + " is not supported");
-            return;
+        if (!isValidFieldDefinition(fieldDefinition, expression.getField())) {
+            return expression;
         }
 
-        if (fieldDefinition.getOperators().stream()
-                .noneMatch(operator -> operator.getName().equals(expression.getOperator()))) {
-            setValidationError("Operator " + expression.getOperator() + " is not supported for field " +
-                    expression.getField());
-            return;
+        if (!isValidOperator(fieldDefinition, expression.getOperator())) {
+            return expression;
         }
 
-        if (!fieldDefinition.getValue().getValueType().name().equals(expression.getValue().getType().name())) {
-            setValidationError("Value type " + expression.getValue().getType().name() + " is not supported for field "
-                    + expression.getField());
+        Value resolvedValue = validateAndResolveValue(fieldDefinition, expression.getValue());
+        if (isError) {
+            return expression;
         }
 
-        if (fieldDefinition.getValue() instanceof OptionsInputValue &&
-                ((OptionsInputValue) fieldDefinition.getValue()).getValues().stream().noneMatch(
-                        optionsValue -> optionsValue.getName().equals(expression.getValue().getFieldValue()))) {
-            setValidationError("Value " + expression.getValue().getFieldValue() + " is not supported for field " +
-                    expression.getField());
+        Expression resolvedExpression = new Expression.Builder()
+                .field(expression.getField())
+                .operator(expression.getOperator())
+                .value(resolvedValue)
+                .build();
+
+        if (!isValidOptionsInputValue(fieldDefinition, expression.getValue().getFieldValue())) {
```

Comment:
Is there any reason to do this validation after building the resolvedExpression.

Reply:
It needs to be removed. Good catch

Reply:
Fixed